File: lib/code_gen/targets/wasm/wasm_block.rb
# typed: strict
# frozen_string_literal: true

require "sorbet-runtime"
require_relative "../../../analysis/bb"

module Lilac
  module CodeGen
    module Targets
      module Wasm
        # A basic WasmBlock.
        class WasmBlock
          extend T::Sig

          include CodeGen::Targets::Wasm

          sig { returns(Analysis::BB) }
          attr_reader :bb

          sig { returns(T.nilable(WasmBlock)) }
          attr_accessor :next_block

          sig do
            params(bb: Analysis::BB, next_block: T.nilable(WasmBlock)).void
          end
          def initialize(bb, next_block: nil)
            @bb = bb
            @next_block = next_block
          end
        end

        # A WasmBlock that represents a true/false conditional.
        class WasmIfBlock < WasmBlock
          extend T::Sig

          sig { returns(T.nilable(WasmBlock)) }
          attr_accessor :true_branch

          sig { returns(T.nilable(WasmBlock)) }
          attr_accessor :false_branch

          sig do
            params(bb: Analysis::BB,
                   true_branch: T.nilable(WasmBlock),
                   false_branch: T.nilable(WasmBlock)).void
          end
          def initialize(bb, true_branch: nil, false_branch: nil)
            @bb = bb
            @true_branch = true_branch
            @false_branch = false_branch
          end
        end

        # A WasmBlock that represents a loop header.
        class WasmLoopBlock < WasmBlock
          extend T::Sig

          sig { returns(T.nilable(WasmBlock)) }
          attr_accessor :inner

          sig { params(bb: Analysis::BB, inner: T.nilable(WasmBlock)).void }
          def initialize(bb, inner: nil)
            @bb = bb
            @inner = inner
          end
        end
      end
    end
  end
end

File: Lilac/src/CodeGen/Targets/Wasm/WasmBlock.cs
using Lilac.Analysis;

namespace Lilac.CodeGen.Targets.Wasm;

public class WasmBlock {
  public BB BB { get; }
  public WasmBlock? NextBlock { get; set; }

  public WasmBlock(BB bb, WasmBlock? nextBlock = null) {
    BB = bb;
    NextBlock = nextBlock;
  }
}

public class WasmIfBlock : WasmBlock {
  public WasmBlock? TrueBranch { get; set; }
  public WasmBlock? FalseBranch { get; set; }

  public WasmIfBlock(BB bb,
                     WasmBlock? trueBranch = null,
                     WasmBlock? falseBranch = null)
    : base(bb) {
    TrueBranch = trueBranch;
    FalseBranch = falseBranch;
  }
}

public class WasmLoopBlock : WasmBlock {
  public WasmBlock? Inner { get; set; }

  public WasmLoopBlock(BB bb, WasmBlock? inner = null)
    : base(bb) {
    Inner = inner;
  }
}